### DISCUSSION TITLE
Bring the middleware functionality back into Payload 3.0

### DISCUSSION THREAD

**Discussion by r1tsuu** [payloadcms/payload#7588] [Feature Requests & Ideas]
Something that I haven't seen discussed, but I think we should be able to apply middlewares just like we did with 2.0 and Express.

Examples of the usage:
Logging, analytics, caching, custom endpoints specific authentication / context modification (DRY)

Express `next` function _could_ be replaced with this logic:
1. If we didn't return `Response` from our middleware, the next middleware / actual handler will be executed with passed `req` through. It would be also nice to be able here to modify headers of the future response.
2. In case if we returned `Response` - exit the execution context
3. In case of an error within the middleware execution, catch it in the same way as we currently do in our endpoints.

**Comment by JeffreyArts** [comment]
Could you please provide some context on how the 3.0 version operates different from Payload 2.0? I have no hands-on experience (yet) with version 3.0, but your remark sounds valuable to me. Removing the option to add middleware sounds like a step towards blackboxing, which sounds contradictory to Payloads (publicly communicated) philosophy/vision.

**Reply by r1tsuu** [reply]
There's no `middleware` option in the config anymore, that's it. What's different - instead of Express we use Next.js now for the HTTP API

**Reply by JeffreyArts** [reply]
Where can I learn more on the inclusion of NextJS into Payload? I just have written a plugin that is dependent on the opportunity to add middleware to Express. Also, by not including express at all, it would be able to write micro services that could be run without any visual interface, which I found rather powerful. But a tighter integration with Next.js sounds like every Payload instance will automatically fire up a server instance, which would not be desirable. I might be wrong on this (since I haven't read into this subject) but it raises some flags for me. Thank you for bringing this to my attention :)

**Reply by r1tsuu** [reply]
Actually Payload 3 doesnt require Next.js for the Local API to be working, it's just for the admin panel / REST / GraphQL. You have abillity to just use the Payload instance in your non Next.js app

**Comment by nachomglz** [comment]
Has this discussion moved forward? I'm now struggling with the same thing. I tried adding the Next.js `middleware.ts` file but it doesn't have access to the `payload` object. Is there a workaround for this? I need to access the DB for custom authentication purposes as you mentioned @r1tsuu

**Reply by kfiross** [reply]
@r1tsuu @nachomglz maybe it's better to use express for it?

**Comment by kfiross** [comment]
@r1tsuu @DracoBlue @nachomglz @JeffreyArts 
I'm worked on a nice solution to get the same functionallity as in next, stay tuned!

https://www.npmjs.com/package/payload-express-middleware
